kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionImpl.java

```diff
@@ -145,6 +152,16 @@ public TransactionCommitResult commit(Engine engine, CloseableIterable<Row> data
       CommitInfo attemptCommitInfo = generateCommitAction(engine);
       updateMetadataWithICTIfRequired(
           engine, attemptCommitInfo.getInCommitTimestamp(), readSnapshot.getVersion(engine));
+
+      // For Row Tracking, we assign base row IDs to all AddFiles inside dataActions that
+      // do not have it yet. If the high water mark has changed, we also emit a
+      // DomainMetadata action with the new high water mark.
+      RowIDAssignmentResult rowIDAssignmentResult =
+          RowTracking.assignBaseRowId(protocol, readSnapshot, FULL_SCHEMA, dataActions);
+      dataActions = rowIDAssignmentResult.getDataActions();
+      domainMetadataIter =
+          domainMetadataIter.combine(rowIDAssignmentResult.getDomainMetadatasIter());
```

Comment:
This is done in this way because I remember we discussed that we want to keep `dataActions` only for data-related actions (`AddFile`, `RemoveFile`). If it’s acceptable to include `DomainMetadata` actions in `dataActions`, we could simplify this by appending potential `DomainMetadata` actions directly to `dataActions`.